docs/04-reference/winglang-spec.md

```diff
@@ -172,45 +172,287 @@ inflight (arg1: <type1>, arg2: <type2>, ...) => <type>
+
+```js
+let p = Json { first: "hello", last: "world", another_field: "ignored" };
+let c = Contact.parse(p);
+assert(c.first == "hello");
+assert(c.last == "world");
+// `c.another_field` is not a thing
+```
+
+This can be disabled using `partial: false` (P2):
+
+```js
+Contact.parse(Json { first: "hello", last: "world", another_field: "ignored" }, partial: false);
+// RUNTIME ERROR: cannot parse Contact due to extraneous field "another_field"
+```
+
+Structs have a `schema` static method which returns a `JsonSchema` object (P2):
+
+```js
+let schema = Contact.schema();
+schema.validate(j);
+```
+
+##### 1.1.4.3. MutJson
+
+To define a mutable JSON container, use the `MutJson` type:
+
+```js
+let my_obj = MutJson { hello: "dear" };
+```
+
+Now you can mutate the contents by assigning values:
+
+```js
+let foo_num = 123;
+my_obj.world = "world";
+my_obj.dang = [1,2,3,4];
+my_obj.sub_object = {};
+my_obj.sub_object.arr = [1,"hello","world"];
+my_obj.foo = foo_num;
+```
+
+To delete a key from an object, use the `Json.delete()` method:
+
+```js
+let my_obj = MutJson { hello: 123, world: 555 };
+Json.delete(my_obj, "world");
+
+let immut_obj = Json { hello: 123 };
+Json.delete(immut_obj, "hello");
+//          ^^^^^^^^^ expected `JsonMut`
+```
+
+To modify a Json array, you will need to parse it into a native `MutArray` and then modify it. This
+implies that at the moment, it is not possible to mutate heterogenous JSON arrays:
+
+```js
+let j1 = MutJson { hello: [1,2,3,4] };
+let a1 = MutArray<num>.parse_json(j1);
+a1.push(5);
+
+j1.hello = a1;
+```
+
+> We will need to revisit this as we progress if this is a major use case.
+
+##### 1.1.4.4 Formatting
+
+The `Json.format(j: Json): str` static method can be used to serialize a `Json` as a string:
+
+```js
+assert(Json.format(json_string) == "\"hello\"");
+assert(Json.format(json_obj) == "{\"boom\":123}");
+assert(Json.format(json_mut_obj, indent: 2) == "{\n\"hello\": 123,\n"  \"world\": [\n    1,\n    2,\n    3\n  ],\n  \"boom\": {\n    \"hello\": 1233\n  }\n}");
+```
+
+##### 1.1.4.5 Equality, diff and patch
+
+The `Json.equals(lhs: Json, rhs: Json): bool` static method can be used to determine if two
+values are equal (recursively comparing arrays and objects):
+
+```js
+assert(Json.equals(json_string, Json "hello"));
+assert(Json.equals(json_obj, { boom: [ 1, 2, 3 ] }));
+assert(!Json.equals(Json { hello: [ 1, 2, 3 ] }, Json { hello: [ 1, 2 ] }));
+```
+
+The `Json.diff(lhs: Json, rhs: Json): JsonPatch` static method can be used to calculate the deep
+difference between two JSON values. It returns a list of differences in
+[json-patch](https://jsonpatch.com/) format.
+
+```js
+let j1 = Json {
+  baz: "qux",
+  foo: "bar"
+};
+
+let j2 = Json {
+  baz: "boo",
+  hello: ["world"]
+};
+
+assert(Json.diff(j1, j2) = [
+  { op: JsonPatch.REPLACE, path: "/baz", value: "boo" },
+  { op: JsonPatch.ADD, path: "/hello", value: ["world"] },
+  { op: JsonPatch.REMOVE, path: "/foo" }
+]);
```

Comment:
How is the list of returned operations ordered? Or should `Json.diff` return a `Set` of operations?

Reply:
I think this can't really be a set. The order can affect the result. I'll leave it for json-diff to take care of .